### PR TITLE
ASoC: SOF: loader: Add debug box region

### PR DESCRIPTION
**Pull request by IulianOlaru249**
This patch adds an IPC initiated debug box region in the snd_sof_dev
structure, defined in soc/sof/sof-priv.h. It is initialized at loading,
in the sof_get_windows funtion from soc/sof/loader.c, in a similar manner
with the stream box and host box.

This region is useful because the firmware will put an error message
here so the kernel can read it in case of a dsp ops.

Signed-off-by: Iulian Olaru <iulianolaru249@yahoo.com>

### CODE REVIEW

**Comment by IulianOlaru249** [comment]
@paulstelian97 this is part of a series of patches that will add debug support on imx platforms. Hence, the do not merge comment. The goal of this patch is to have a distinct memory area from where the kernel will read panic codes placed by the firmware. 
Difrent messages can also get here but the debug codes will be handled soon, in a difrent patch.

**Comment by paulstelian97** [comment]
> @paulstelian97 this is part of a series of patches that will add debug support on imx platforms. Hence, the do not merge comment. The goal of this patch is to have a distinct memory area from where the kernel will read panic codes placed by the firmware.
> Difrent messages can also get here but the debug codes will be handled soon, in a difrent patch.

I shall ping @plbossart @lgirdwood to see if there is any ABI change that should happen to introduce this memory region.

**Comment by IulianOlaru249** [comment]
@dbaluta you might also want to take a look.

File: sound/soc/sof/sof-priv.h [hunk 1]
```diff
@@ -386,6 +386,7 @@ struct snd_sof_dev {
 	struct snd_sof_mailbox dsp_box;		/* DSP initiated IPC */
 	struct snd_sof_mailbox host_box;	/* Host initiated IPC */
 	struct snd_sof_mailbox stream_box;	/* Stream position update */
+	struct snd_sof_mailbox debug_box;	/* Debug initiated IPC */
 	struct snd_sof_ipc_msg *msg;
 	int ipc_irq;
```
Comment (dbaluta) [inline sound/soc/sof/sof-priv.h:391]: Comment here can be enhanced. 'Debug initiated IPC' doesn't make sense because Debug is not an entity who can do stuff like DSP or Host as we see in the above comments for dsp_box and host_box.
We can either remove this comment or say ``` /* debug info updates */```

File: sound/soc/sof/loader.c [hunk 2]
```diff
@@ -310,6 +310,8 @@ static void sof_get_windows(struct snd_sof_dev *sdev)
 	u32 outbox_size = 0;
```
Comment (dbaluta) [inline sound/soc/sof/loader.c:310]: Commit subject prefix should be 'ASoC: SOF: loader' we always use SOF instead of Sof. Also please remove dot at the end of commit subject.
We should also explain better why do we need this.

**Comment by dbaluta** [comment]
@plbossart @lgirdwood @ranj063 @paulstelian97 @IulianOlaru249. This change is needed because our Messaging Unit (equivalent of SHIM layer for Intel) doesn't have an easy way of storing information together with notification interrupts.

On Intel, before handling a message we can check if an actual panic had taken place by this:

```
static irqreturn_t byt_irq_thread(int irq, void *context) {
»       ipcd = snd_sof_dsp_read64(sdev, BYT_DSP_BAR, SHIM_IPCD);
»       »       if ((ipcd & SOF_IPC_PANIC_MAGIC_MASK) == SOF_IPC_PANIC_MAGIC) {

```
IPCD register is a 64 bit register with bits 63 and 62 used for notification but with additional 62 bits for sending information. 

We don't have this easily with our Messaging Unit so we need another way of sending out of band information. I think if Intel folks agree we can make the debug box generic to all platforms.

But we can go now with IMX implementation given that that debug box area is not used for other stuff. Even if it is used for other stuff, we can reserve 4 bytes for sending our panic status from DSP.

We need to make sure that this isn't somehow breaking the ABI.

File: sound/soc/sof/loader.c [hunk 3]
```diff
@@ -310,6 +310,8 @@ static void sof_get_windows(struct snd_sof_dev *sdev)
...
 	u32 stream_size = 0;
 	u32 inbox_size = 0;
+	u32 debug_size = 0;
```
Comment (ranj063) [inline sound/soc/sof/loader.c:313]: is there a typo in the commit message? "in case of dsp ops" or "in case of dsp oops"?

**Comment by dbaluta** [comment]
SOFCI TEST

**Comment by dbaluta** [comment]
SOFCI TEST

**Comment by dbaluta** [comment]
@plbossart @ranj063 @singalsu  Any idea what's with the checkpatch.pl error:

```
+scripts/checkpatch.pl --strict --codespell -g 9724bf01bb2ba36d7a7f2e0e2c544093db19c140..a316ce5fdd7a304e59555d5faacc8ae51e8e112f

----------------------------------------

Commit 1da177e4c3f4 ("Linux-2.6.12-rc2")

----------------------------------------

Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.

Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.

Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.

Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.

Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.

Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.

Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.

Use of uninitialized value $typo_fix in concatenation (.) or string at scripts/checkpatch.pl line 2968.

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

The build has been terminated
```

```

**Comment by plbossart** [comment]
> @plbossart @ranj063 @singalsu Any idea what's with the checkpatch.pl error:
> 
> ```
> +scripts/checkpatch.pl --strict --codespell -g 9724bf01bb2ba36d7a7f2e0e2c544093db19c140..a316ce5fdd7a304e59555d5faacc8ae51e8e112f
> 
> ----------------------------------------
> 
> Commit 1da177e4c3f4 ("Linux-2.6.12-rc2")
> 
> ----------------------------------------
> 
> Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.
> 
> Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.
> 
> Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.
> 
> Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.
> 
> Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.
> 
> Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.
> 
> Use of uninitialized value $typo_fix in ucfirst at scripts/checkpatch.pl line 2964.
> 
> Use of uninitialized value $typo_fix in concatenation (.) or string at scripts/checkpatch.pl line 2968.
> 
> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
> 
> Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
> 
> The build has been terminated
> ```

no idea, this looks like either a Travis bug or a scripting bug. @marc-hb @xiulipan any ideas?

**Comment by marc-hb** [comment]
These checkpatch error messages are what you get when you run checkpatch on 1da177e4c3f4 which is the very first commit in the entire git history of the kernel.

Looking at https://travis-ci.org/github/thesofproject/linux/jobs/716489255 (please always save these URLs in comments because github loses them), I don't understand why checkpatch is run on this commit. Looking at the Travis log the checkpatch arguments all look correct and never referring to commit 1da177e4c3f4

**Comment by marc-hb** [comment]
I think I found a clue thanks to the extra logging I recently added to `.travis.yml` in commit 1a7597ab8c89: checkpatch is probably confused by the shallow clone we're telling Travis to do. I couldn't reproduce this exactly but I did manage to confuse checkpatch with a shallow clone too.

Because of a recent merge, this PR is now based thousands of commits behind the latest HEAD  which when combined with a shallow clone makes things "interesting" from a git perspective. In fact when you look at the timings on the right side of https://travis-ci.org/github/thesofproject/linux/jobs/716489255 you can see the shallow cloning speed optimization is made completely pointless because the next git fetch (of this PR) takes more than 10 minutes anyway immediately after.

@plbossart is the shallow cloning always that useless or is it only because of this recent, very big upstream merge you did? It would be just a one-line change to turn it off in .travis.yml (or to extend it to 50, the maximum)

**Comment by dbaluta** [comment]
@plbossart @lgirdwood this is good to merge. checkpatch.pl problem is not related to this patch

**Comment by plbossart** [comment]
````
 git clone --depth=20 https://github.com/thesofproject/linux.git thesofproject/linux
````

It's unclear to me what branch is used here? I think it fetches the master branch, which is completely useless indeed. Well, unless it's a required step and the --depth is a work-around to avoid using it really.

**Comment by paulstelian97** [comment]
@plbossart It will fetch the default branch which for this repo was set to topic/sof-dev instead of master. (also why did quote reply not work correctly?...)

**Comment by marc-hb** [comment]
> I think it fetches the master branch, which is completely useless indeed. 

As explained by @paulstelian97 , it doesn't fetch the master branch. `git clone` fetches the default branch.

So, good news: the`git fetch origin +refs/pull/2341/merge` in https://travis-ci.org/github/thesofproject/linux/jobs/716489255 was 11 minutes long ONLY because the PR was thousands of commits behind the shallow `topic/sof-dev` default branch. To verify this I looked at another, random but recent PR that is NOT thousands of commits behind and the equivalent `git fetch +refs/pull/2411/merge` took 1 second: https://travis-ci.org/github/thesofproject/linux/jobs/722753785 (this entire job took 2 minutes!)

Warning: Travis timings in the right margin are VERY misleading, for instance the initial `git clone` in https://travis-ci.org/github/thesofproject/linux/jobs/716489255 looks like it has a 71s duration. Then you click on the triangle to unfold it and you discover this 11 minutes long git fetch inside a 71s long git clone step...

**Comment by marc-hb** [comment]
tl;dr: checkpatch should be better tested with shallow clones.

It should especially be tested with clones with "gaps" (combination of full old clone + recent shallow clone)

It will never make up missing commits but it should at least provide less confusing error messages.
